### PR TITLE
feat(mailbox): persistent bus subscriptions + glob topic patterns

### PR DESCRIPTION
**Pull request by sam-phinizy**
## Summary
- Adds persistent bus subscriptions: an alias subscribes to a topic glob (`repo-a.*`, `**.completed`) and matched bus events land in the subscriber's mailbox. `roux mailbox read` returns them, the unread badge bumps via `mailbox-event`, and the subscriber can mark-read / ack like any other mail. Subscriptions persist to `subscriptions.json`.
- MQTT-style globbing: `*` matches one segment, `**` matches many. Pattern segments use the same alphabet as topics (lowercase, digits, hyphens). Partial-segment globs (`foo*`) are rejected at validation.
- Closes the **persistent bus subscriptions** and **glob topic patterns** items from #161.

## Stack
This PR is **stacked on top of #162** (\`feature/improve-mailbox\` → \`main\`). The base of this PR is \`feature/improve-mailbox\` so the diff shows only the subscription work; once #162 lands, this auto-rebases onto \`main\`. **#164** (mailbox-watch) is stacked on top of this one.

Review order: **#162 → #163 → #164**.

## Surface
- Core: \`topic_glob\` matcher (22 tests); \`BusSubscription\` model; \`MailboxEvent::TopicDelivered\` variant.
- Backend: \`subscriptions/\` module mirroring \`aliases/\` (store / persistence / manager). Mailbox store gains pattern-aware \`list_for_recipient\` / \`recipient_owns\` / \`unread_count\` so subscribers see and ack topic events without losing the existing addressed-mail gatekeeping.
- CLI: \`roux bus subscribe '<pattern>' [--alias <name>] [--project <id>]\`, \`roux bus unsubscribe <id>\`, \`roux bus subscriptions [--alias <name>]\`. \`--alias\` defaults to the calling pane's auto-claimed alias.
- MCP: \`roux_bus_subscribe\` / \`roux_bus_unsubscribe\` / \`roux_bus_subscriptions\` tools.
- Tauri commands: \`subscriptions_list\` / \`subscriptions_create\` / \`subscriptions_delete\`.
- UI: Subs tab in the Mailbox panel — list, create, delete, with live updates via the new \`subscription-event\` Tauri channel.
- Persistence: versioned \`subscriptions.json\` envelope (mirrors \`aliases.json\`).

## Test plan
- [x] \`cargo test --bin roux\` — 478 pass
- [x] \`cargo test --lib --workspace\` — 385 pass across crates
- [x] \`npm run test\` — 1003 pass (4 new for the subscription store)
- [x] \`npm run check\` — 0 errors
- [x] \`cargo clippy --bin roux --tests\` — clean
- [ ] Manual end-to-end (per the design doc):
  - Subscribe \`auditor\` to \`**.completed\`; publish from another pane; confirm unread badge ticks and \`roux mailbox read --ack\` from auditor's pane returns the event.
  - Subscribe with a project scope; publish without that scope; confirm no match.
  - Delete via the Subs tab; confirm next publish is silent.

## Out of scope (explicit)
- \`mailbox-watch\` push CLI — addressed in **#164**, stacked on top of this PR.
- Group aliases.
- Subscription rate-limiting / pause-resume.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

- Adds persistent bus subscriptions with MQTT-style glob matching (`*` = one segment, `**` = zero-or-more). The memoized `O(P×T)` matcher, duplicate-triple guard, versioned persistence, and cross-project authorization fix (`patterns_for_event` scoped to the event's own project for `mark_read`/`ack`) are all correct and well-tested.
- The previous listener-leak issue is resolved: `startSubscriptionEventListenerWithCleanup` returns `stopSubscriptionEventListener` and is properly pushed to `tauriUnlisteners` in `App.svelte`.
- One P2 in `SubscriptionManager::unsubscribe`: when `persist()` fails and a concurrent `subscribe` has raced to re-add the same triple, the rollback `store.add(removed)` silently fails, leaving memory and disk transiently out of sync until the next restart.

<details><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge; the single P2 is an edge-case race on a rare error path with no data loss beyond the current session.

Only P2 findings present — the silent rollback failure in `unsubscribe` requires a concurrent subscribe racing a persist failure, which is extremely unlikely in a desktop app. All core paths (glob matching, project-scoped authorization, persist-before-emit, listener cleanup) are correct and well-tested.

src-tauri/src/subscriptions/manager.rs — the `unsubscribe` rollback at line 136–139.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| crates/roux-core/src/topic_glob.rs | Well-implemented MQTT-style glob matcher with memoized O(P×T) matching; 22 tests cover edge cases including adversarial multi-`**` patterns. |
| src-tauri/src/subscriptions/store.rs | Clean in-memory store with correct duplicate-triple prevention and project-scope matching; `patterns_for_alias` correctly includes only scope-compatible patterns. |
| src-tauri/src/subscriptions/manager.rs | Correct persist-before-emit pattern mirrors alias manager; minor silent rollback issue in `unsubscribe` if concurrent subscribe races with failed persist. |
| src-tauri/src/subscriptions/persistence.rs | Versioned JSON envelope with backcompat bare-array fallback; mirrors aliases persistence pattern faithfully. |
| src-tauri/src/mailbox/store.rs | Subscription-aware visibility and ownership correctly layered on top of existing addressed-mail semantics; `recipient_owns` None-arm (any recipient can track state on topic-only events) is intentional and documented. |
| src-tauri/src/mailbox/manager.rs | `patterns_for_event` (not `patterns_for(Any)`) is used for `mark_read`/`ack` ownership so cross-project authorization is correctly blocked; the `TopicDelivered` fan-out on `post` is clean. |
| src-tauri/src/commands/subscriptions.rs | Thin Tauri command adapter; delegates all logic to `SubscriptionManager`; consistent with repo convention. |
| src-tauri/src/socket.rs | `resolve_subscriber_alias` correctly disambiguates explicit vs pane-context aliases and errors on multi-alias panes; new bus-subscribe/unsubscribe/subscriptions handlers are consistent with existing socket command style. |
| src/lib/stores/subscriptions.ts | Previous listener-leak issue resolved: `startSubscriptionEventListenerWithCleanup` returns `stopSubscriptionEventListener` and is pushed to `tauriUnlisteners` in App.svelte. |
| src/lib/components/MailboxPanel.svelte | Subs tab wired correctly; global subscription list is intentional; alias pre-fill on tab switch is a nice UX touch. |
| crates/roux-core/src/models/subscription.rs | New `BusSubscription` and `BusSubscriptionEvent` models are well-structured and match the TypeScript types exposed via `tauri.ts`. |
| src-tauri/src/mcp.rs | Three new MCP tools (`roux_bus_subscribe`, `roux_bus_unsubscribe`, `roux_bus_subscriptions`) route through the socket, consistent with all other MCP tools in the file. |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant CLI as CLI / MCP
    participant Socket as Socket Handler
    participant SubMgr as SubscriptionManager
    participant SubStore as SubscriptionStore
    participant Disk as subscriptions.json
    participant Tauri as Tauri Events

    CLI->>Socket: "bus-subscribe {pattern, alias}"
    Socket->>Socket: resolve_subscriber_alias()
    Socket->>SubMgr: subscribe(alias, pattern, project_id)
    SubMgr->>SubMgr: validate_alias_name + validate_topic_pattern
    SubMgr->>SubStore: add(subscription) — duplicate-triple check
    SubStore-->>SubMgr: Ok(inserted)
    SubMgr->>Disk: save_to_path (persist before emit)
    Disk-->>SubMgr: Ok
    SubMgr->>Tauri: emit(subscription-event, Created)
    Tauri-->>CLI: Ok(subscription)

    note over CLI,Tauri: Later: bus publish triggers delivery

    CLI->>Socket: "bus-publish {topic, body}"
    Socket->>SubMgr: matching_topic(topic, project_id)
    SubMgr-->>Socket: [BusSubscription, ...]
    Socket->>SubStore: post event to mailbox
    loop for each matching subscription
        Socket->>Tauri: emit(mailbox-event, TopicDelivered)
    end
```

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Asrc-tauri%2Fsrc%2Fsubscriptions%2Fmanager.rs%3A136-139%0A**Silent%20rollback%20failure%20in%20%60unsubscribe%60**%0A%0AWhen%20%60persist%28%29%60%20fails%20after%20the%20in-memory%20removal%2C%20the%20rollback%20calls%20%60store.add%28removed%29%60%20but%20silently%20discards%20its%20result%20with%20%60let%20_%60.%20In%20the%20narrow%20window%20between%20the%20first%20mutex%20release%20%28after%20%60store.remove%60%29%20and%20the%20rollback%20re-lock%2C%20a%20concurrent%20%60subscribe%60%20call%20with%20the%20same%20%60%28alias%2C%20pattern%2C%20project_id%29%60%20triple%20could%20have%20already%20re-added%20the%20entry.%20The%20%60add%60%20inside%20the%20rollback%20will%20then%20return%20%60Err%28Duplicate%29%60%20%E2%80%94%20that%20error%20is%20dropped%2C%20so%20the%20rollback%20is%20a%20no-op.%0A%0AThe%20end%20state%3A%20memory%20has%20the%20concurrent%20subscriber's%20new%20entry%20%28not%20the%20original%29%2C%20while%20disk%20still%20contains%20the%20original.%20From%20the%20user's%20perspective%20the%20subscription%20looks%20gone%3B%20on%20the%20next%20restart%20it%20reappears%20from%20disk%2C%20at%20which%20point%20the%20user%20could%20have%20two%20logically%20identical%20entries.%0A%0A%60%60%60rust%0A%2F%2F%20Instead%20of%20silently%20ignoring%20the%20failure%2C%20log%20it%20so%20it's%20diagnosable.%0Aif%20let%20Err%28add_err%29%20%3D%20store.add%28removed%29%20%7B%0A%20%20%20%20eprintln!%28%22%5Broux%5D%20unsubscribe%20rollback%20failed%20%28entry%20not%20restored%29%3A%20%7Badd_err%7D%22%29%3B%0A%7D%0A%60%60%60%0A%0A&repo=phin-tech%2Froux&pr=163&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=2">Fix All in Claude Code</picture></a> <a href="https://chatgpt.com/codex/deeplink?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22phin-tech%2Froux%22%20on%20the%20existing%20branch%20%22feature%2Fbus-subscriptions%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22feature%2Fbus-subscriptions%22.%0A%0AFix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Asrc-tauri%2Fsrc%2Fsubscriptions%2Fmanager.rs%3A136-139%0A**Silent%20rollback%20failure%20in%20%60unsubscribe%60**%0A%0AWhen%20%60persist%28%29%60%20fails%20after%20the%20in-memory%20removal%2C%20the%20rollback%20calls%20%60store.add%28removed%29%60%20but%20silently%20discards%20its%20result%20with%20%60let%20_%60.%20In%20the%20narrow%20window%20between%20the%20first%20mutex%20release%20%28after%20%60store.remove%60%29%20and%20the%20rollback%20re-lock%2C%20a%20concurrent%20%60subscribe%60%20call%20with%20the%20same%20%60%28alias%2C%20pattern%2C%20project_id%29%60%20triple%20could%20have%20already%20re-added%20the%20entry.%20The%20%60add%60%20inside%20the%20rollback%20will%20then%20return%20%60Err%28Duplicate%29%60%20%E2%80%94%20that%20error%20is%20dropped%2C%20so%20the%20rollback%20is%20a%20no-op.%0A%0AThe%20end%20state%3A%20memory%20has%20the%20concurrent%20subscriber's%20new%20entry%20%28not%20the%20original%29%2C%20while%20disk%20still%20contains%20the%20original.%20From%20the%20user's%20perspective%20the%20subscription%20looks%20gone%3B%20on%20the%20next%20restart%20it%20reappears%20from%20disk%2C%20at%20which%20point%20the%20user%20could%20have%20two%20logically%20identical%20entries.%0A%0A%60%60%60rust%0A%2F%2F%20Instead%20of%20silently%20ignoring%20the%20failure%2C%20log%20it%20so%20it's%20diagnosable.%0Aif%20let%20Err%28add_err%29%20%3D%20store.add%28removed%29%20%7B%0A%20%20%20%20eprintln!%28%22%5Broux%5D%20unsubscribe%20rollback%20failed%20%28entry%20not%20restored%29%3A%20%7Badd_err%7D%22%29%3B%0A%7D%0A%60%60%60%0A%0A"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=2">Fix All in Codex</picture></a>

<sub>Reviews (3): Last reviewed commit: ["fix: address PR #163 review findings"](https://github.com/phin-tech/roux/commit/f0891a742732c2a44fb98fea1bc8df8aa58bec55) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31439168)</sub>

### CODE REVIEW

File: src-tauri/src/mailbox/manager.rs [hunk 1]
```diff
@@ -52,14 +58,48 @@ impl MailboxManager {
...
+            // hold regardless of project — visibility broadens with the
+            // request scope.
+            ProjectFilter::Any => {
+                subs.for_alias(recipient, ProjectFilter::Any)
+                    .into_iter()
+                    .map(|s| s.pattern)
+                    .collect()
+            }
+            ProjectFilter::Exact(scope) => subs.patterns_for_alias(recipient, scope),
+        }
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src-tauri/src/mailbox/manager.rs:100]: Fixed in f0891a7. Manager now passes per-event patterns to the store: `patterns_for_event` looks up `subs.patterns_for_alias(recipient, event.project_id)` so a p2-scoped sub can't authorize p1 ownership. Added regression test `cross_project_subscription_cannot_authorize_other_project_events`.

File: src-tauri/src/mailbox/manager.rs [hunk 2]
```diff
@@ -52,14 +58,48 @@ impl MailboxManager {
...
     }
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src-tauri/src/mailbox/manager.rs:102]: Fixed in f0891a7 — same finding as coderabbit's #3213115376. `patterns_for_event` now scopes to the event's own project before passing to `recipient_owns`. Regression test added.

File: src-tauri/src/socket.rs [hunk 3]
```diff
@@ -1933,6 +1936,65 @@ async fn handle_bus_tail(req: Request, app: &tauri::AppHandle) -> Response {
...
+        "no --alias given and no pane context available; pass --alias <name>".to_string()
+    })?;
+    let mut held = state.alias_manager.find_for_pane(pane_id);
+    held.sort_by_key(|a| a.created_at);
+    held.into_iter()
+        .next()
+        .map(|a| a.alias)
+        .ok_or_else(|| {
+            "no --alias given and the calling pane holds no alias; pass --alias <name>".to_string()
+        })
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src-tauri/src/socket.rs]: Fixed in f0891a7. `resolve_subscriber_alias` now returns `'caller pane holds multiple aliases (...); pass --alias <name>'` instead of silently picking by `created_at`.

[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src-tauri/src/socket.rs]: Fixed in f0891a7 — same as coderabbit's #3213115378. Disambiguation error replaces silent pick.

File: src-tauri/src/subscriptions/manager.rs [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        let Some(path) = self.persistence_path.as_ref() else {
+            return;
+        };
+        let store = self.inner.lock().expect("subscription store poisoned");
+        if let Err(e) = save_to_path(store.entries(), path.as_path()) {
+            eprintln!(
+                "[roux] subscription persistence failed at {}: {e}",
+                path.display(),
+            );
+        }
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src-tauri/src/subscriptions/manager.rs]: Fixed in f0891a7. `persist()` now returns `io::Result`; `subscribe()` rolls back the in-memory insert and returns `SubscribeError::Persist`. `unsubscribe()` return type became `Result<bool, UnsubscribeError>` and re-inserts the row on save failure. Tauri command + socket handler propagate the error.

File: crates/roux-core/src/models/subscription.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    pub id: String,
+    /// Canonical alias receiving deliveries. Need not be currently bound
+    /// to a pane — queued mail accrues for whoever claims it next.
+    pub alias: String,
+    /// Validated glob pattern. Empty / invalid patterns are rejected at
+    /// the manager boundary.
+    pub pattern: String,
+    /// Project scope. `None` = global. A subscription only matches
+    /// events whose `project_id` is `None` or equals this scope.
+    #[serde(default)]
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply crates/roux-core/src/models/subscription.rs]: Fixed in f0891a7. Doc now states the asymmetry: `None` = global (matches any scope), `Some(p)` only matches events with `project_id == p`.

File: crates/roux-core/src/models/subscription.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    /// guessing on alias+pattern equality.
+    pub id: String,
+    /// Canonical alias receiving deliveries. Need not be currently bound
+    /// to a pane — queued mail accrues for whoever claims it next.
+    pub alias: String,
+    /// Validated glob pattern. Empty / invalid patterns are rejected at
+    /// the manager boundary.
+    pub pattern: String,
+    /// Project scope. `None` = global. A subscription only matches
+    /// events whose `project_id` is `None` or equals this scope.
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply crates/roux-core/src/models/subscription.rs]: Fixed in f0891a7 — same as coderabbit's #3213115370.

File: src-tauri/src/mcp.rs [hunk 7]
```diff
@@ -308,6 +308,32 @@ pub struct BusTailParams {
...
+#[derive(Debug, Deserialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct BusSubscribeParams {
+    /// Glob pattern. `*` matches one segment, `**` matches many.
+    pub pattern: String,
+    /// Alias to bind the subscription to. Defaults to the calling
+    /// pane's alias when omitted.
+    pub alias: Option<String>,
+    pub project_id: Option<String>,
+}
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src-tauri/src/mcp.rs:321]: Fixed in f0891a7. `BusSubscribeParams.alias` is now a required `String` (was `Option<String>`). The CLI fallback to pane alias still works because the CLI handler resolves before the socket call; MCP callers without pane context now get an invalid-params error instead of a silent failure.

File: crates/roux-core/src/topic_glob.rs [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+            if matches_segments(rest_pat, topic) {
+                return true;
+            }
+            // one or more segments: consume one topic segment, retry
+            for i in 1..=topic.len() {
+                if matches_segments(rest_pat, &topic[i..]) {
+                    return true;
+                }
+            }
+            false
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply crates/roux-core/src/topic_glob.rs]: Fixed in f0891a7. Added a `(pi, ti)` memo table to `matches_segments` — worst case is now O(P*T). New test `many_double_stars_over_long_topic_is_fast` asserts 16-segment topic vs 8-`**` pattern under 100ms.

File: crates/roux-core/src/topic_glob.rs [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+        }
+        (Some(_), None) => false,
+        (Some(p), Some(t)) => {
+            let head_ok = *p == "*" || *p == *t;
+            head_ok && matches_segments(&pattern[1..], &topic[1..])
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply crates/roux-core/src/topic_glob.rs]: Fixed in f0891a7 — same as coderabbit's #3213115375. Memoized matcher; perf regression test added.

File: src/lib/stores/subscriptions.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  });
+}
+
+let unlisten: (() => void) | null = null;
+
+/** Wire the global event listener. Idempotent across hot reloads. */
+export async function startSubscriptionEventListener(): Promise<void> {
+  if (unlisten) return;
+  unlisten = await onBusSubscriptionEvent(applySubscriptionEvent);
+}
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src/lib/stores/subscriptions.ts:63]: Fixed in f0891a7. Added an in-flight `listenerInit` promise — subsequent calls await the same registration instead of double-subscribing.

File: src/lib/components/MailboxPanel.svelte [hunk 11]
```diff
@@ -53,7 +58,17 @@
...
   let postError = $state<string | null>(null);
-  let view = $state<"mailbox" | "firehose">("mailbox");
+  let view = $state<"mailbox" | "firehose" | "subscriptions">("mailbox");
+
+  // Subscription compose form state. Shown only when the Subscriptions
+  // tab is active. Pattern is validated server-side; we surface the
+  // backend error inline to keep the form concrete-feedback-driven.
+  let subAlias = $state("");
+  let subPattern = $state("");
+  let subProjectId = $state<string | null>(null);
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src/lib/components/MailboxPanel.svelte]: Fixed in f0891a7 — dropped `subProjectId` state and pass `null` explicitly. Documented in code that scoped subscriptions are CLI/MCP-only for now until we add a proper project picker.

File: src/App.svelte [hunk 12]
```diff
@@ -818,6 +822,8 @@
     // cheap (one list-all + one alias-list + per-alias unread counts);
     // the listener keeps the store in sync without polling.
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src/App.svelte:823]: Fixed in f0891a7. Added `startSubscriptionEventListenerWithCleanup()` that returns the unlisten function; App.svelte now pushes it onto `tauriUnlisteners` alongside the other listeners so hot-reload + shutdown clean up properly.

File: src-tauri/src/subscriptions/persistence.rs [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        Err(_) => return Vec::new(),
+    };
+
+    if let Ok(file) = serde_json::from_str::<SubscriptionsFile>(&content) {
+        return file.data;
+    }
+
+    // Defensive fallback: bare arrays (hand-edited or pre-versioning).
+    serde_json::from_str::<Vec<BusSubscription>>(&content).unwrap_or_default()
+}
```
[the remark this answers is not in the document]
Comment (sam-phinizy) [reply src-tauri/src/subscriptions/persistence.rs:46]: Won't fix in this PR: keeping the silent-recovery behavior matches `aliases.json` and `events.jsonl` loaders. Switching this surface alone would create an inconsistent user experience (one subsystem panics on a corrupt file, the others recover). Tracking as a separate cross-cutting hardening pass — when we revisit, all three persistence layers should adopt typed load errors together so the failure surface is uniform.
